### PR TITLE
Add header and footer report customization

### PR DESCRIPTION
**Pull request by asteriscos**
# Description

This PR is part of White-Labeling Epic https://github.com/wazuh/wazuh-kibana-app/issues/4392.
It adds custom text to the PDF report header and footer.

## Changes
- Add new plugin settings:
  - `customization.reports.header`: customize the footer of the PDF reports. Restricted to `4` lines.
  - `customization.reports.footer`: customize the header of the PDF reports Restricted to `2` lines.
  
| key                                            | description                                                                                            | values             |
| ---                                              | ---                                                                                                          | ---                    |
| `customization.reports.header` | Set the custom header of the PDF reports | String. Maximum lines: 2. If the value is an empty string, the default value will be displayed.   |
| `customization.reports.footer`   | Set the custom footer of the PDF reports | String. Maximum lines: 2. If the value is an empty string, the default value will be displayed.   |

- When the settings contain multiple lines, these are stored as `\n` in the configuration file.
- Add new form input of type `textarea`
- Tests:
  - Validate the setting in the API endpoint to update it
  - `InputForm` component of type `textarea` matches the snapshot

Closes https://github.com/wazuh/wazuh-kibana-app/issues/4426
Base: [feat/settings-validation](https://github.com/wazuh/wazuh-kibana-app/tree/feat/settings-validation)

## Screenshots

image

image

# Tests

## Manual tests

**Scenario**: The `customization.reports.*` settings should appear in the default configuration file
  **Given** there is not a plugin configuration file
  **When** the user starts the plugin platform
  **Then** the configuration file should be generated and includes the `customization.reports.*` settings
  
**Scenario**: The `customization.reports.*` settings should appear in the `Settings/Configuration` section
  **When** the user navigates to `Settings/Configuration`
  **Then** the `customization.reports.*` settings should be there
  
**Scenario**: Set the `customization.reports.header` with a valid value through `Settings/Configuration`.
  **When** the user navigates to `Settings/Configuration` and sets the `customization.reports.header` with a valid value. Maximum 4 lines.
  **Then** the user can save the configuration successfully.
  **And** the configuration file is updated with the configuration.
  
**Scenario**: Set the `customization.reports.footer` with a valid value through `Settings/Configuration`.
  **When** the user navigates to `Settings/Configuration` and sets the `customization.reports.header` with a valid value. Maximum 2 lines.
  **Then** the user can save the configuration successfully.
  **And** the configuration file is updated with the configuration.

**Scenario**: Set the `customization.reports.header` with a valid value through `Settings/Configuration` that includes multiple lines.
  **When** the user navigates to `Settings/Configuration` and sets the `customization.reports.header` with a valid value. Maximum 4 lines.
  **Then** the user can save the configuration successfully.
  **And** the configuration file is updated with the configuration.
  **And** the configuration file is valid and parseable.
  
**Scenario**: Set the `customization.reports.footer` with a valid value through `Settings/Configuration` that includes multiple lines.
  **When** the user navigates to `Settings/Configuration` and sets the `customization.reports.header` with a valid value. Maximum 2 lines.
  **Then** the user can save the configuration successfully.
  **And** the configuration file is updated with the configuration.
  **And** the configuration file is valid and parseable.
  
**Scenario**: Set the `customization.reports.header` with a invalid value through `Settings/Configuration`.
  **When** the user sets the `customization.reports.header` with an invalid value. Maximum 4 lines.
  **Then** a validation error is displayed.
  
**Scenario**: Set the `customization.reports.footer` with a invalid value through `Settings/Configuration`.
  **When** the user sets the `customization.reports.header` with an invalid value. Maximum 2 lines.
  **Then** a validation error is displayed.
  
**Scenario**: Set the `customization.reports.header` with a valid value through an API request.
  **When** the user does an API request to update the `customization.reports.header` setting with a valid value. Maximum 4 lines.
  **Then** the endpoint replies with a successful response.
  **And** the configuration file is updated with the configuration.
  
**Scenario**: Set the `customization.reports.footer` with a valid value through an API request.
  **When** the user does an API request to update the `customization.reports.footer` setting with a valid value. Maximum 2 lines.
  **Then** the endpoint replies with a successful response.
  **And** the configuration file is updated with the configuration.
    
**Scenario**: Set the `customization.reports.header` with an invalid value through an API request.
  **When** the user does an API request to update the `customization.reports.header` setting with an invalid value. Maximum 4 lines.
  **Then** the endpoint replies with a bad request response.
  
**Scenario**: Set the `customization.reports.footer` with an invalid value through an API request.
  **When** the user does an API request to update the `customization.reports.header` setting with an invalid value. Maximum 2 lines.
  **Then** the endpoint replies with a bad request response.
  
**Scenario**: Generate a PDF with a custom report header.
  **Given** the `customization.reports.header` setting has a custom value.
  **When** generates a PDF module report.
  **Then** the custom text should appear in the header.
  
**Scenario**: Generate a PDF with a custom report footer.
  **Given** the `customization.reports.footer` setting has a custom value.
  **When** generates a PDF module report.
  **Then** the custom text should appear in the footer.
  
**Scenario**: Generate a PDF with an empty custom report header.
  **Given** the `customization.reports.header` setting has an empty custom value `""`
  **When** generates a PDF module report.
  **Then** the default text should appear in the header.
  
**Scenario**: Generate a PDF with a custom report footer.
  **Given** the `customization.reports.footer` setting has an empty custom value `""`
  **When** generates a PDF module report.
  **Then** the default text should appear in the footer.
  
  
  ***
  To manually test the configuration endpoint here are 2 scripts to update different configuration values.
  
[pdf-report-conf.zip](https://github.com/wazuh/wazuh-kibana-app/files/9735019/pdf-report-conf.zip)

To use it, first unzip the files in the same folder and allow both _executing the file as a program_. Afterwards you can run the `manual-test-header-footer.sh` script.

You can also use **wz-api-conf.sh** script with custom configuration parameters, like:
```
\- s setting_name
\- v setting_value
```

eg.
`./wz-api-conf.sh -s disabled_roles -v [\"disabled\"]`

#### Output

```
./manual-test-header-footer.sh 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    829      0 --:--:-- --:--:-- --:--:--   827
{"customization.reports.footer":"Custom footer"}
{"data":{"requiresRunningHealthCheck":false,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":false,"updatedConfiguration":{"customization.reports.footer":"Custom footer"}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    896      0 --:--:-- --:--:-- --:--:--   894
{"customization.reports.header":"Custom header"}
{"data":{"requiresRunningHealthCheck":false,"requiresReloadingBrowserTab":false,"requiresRestartingPluginPlatform":false,"updatedConfiguration":{"customization.reports.header":"Custom header"}}}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    906      0 --:--:-- --:--:-- --:--:--   906
{"customization.reports.footer":1111}
{"statusCode":400,"error":"Bad Request","message":"[request body.customization.reports.footer]: expected value of type [string] but got [number]"}  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   398  100   398    0     0    902      0 --:--:-- --:--:-- --:--:--   902
{"customization.reports.header":true}
{"statusCode":400,"error":"Bad Request","message":"[request body.customization.reports.header]: expected value of type [string] but got [boolean]"}
```

## Automatic tests
- `public/components/common/form/index.test.tsx`
- `server/routes/wazuh-utils/wazuh-utils.test.ts`
- `common/plugin-settings.test.ts`
- `server/routes/wazuh-reporting.test.ts`

**Run all PR tests**
```shell
yarn test:jest public/components/common/form/index.test.tsx server/routes/wazuh-utils/wazuh-utils.test.ts common/plugin-settings.test.ts server/routes/wazuh-reporting.test.ts
```

### CODE REVIEW

**Comment by asteriscos** [comment]
### Examples of different logos ratios

**_Before_**
Screenshot from 2022-09-14 17-21-41

**_After_**
More width than height:
Screenshot from 2022-09-14 18-49-40

Equal width and height:
Screenshot from 2022-09-14 18-48-14

***


**_Wazuh Logo for reference_**
image

**Comment by Desvelao** [comment]
I added some manual tests to the initial comment.

**Comment by Desvelao** [comment]
# Changes

- Updated development branch with the last changes of the base branch.

**Comment by Desvelao** [comment]
# Update

- Update the development branch with the last changes of the base branch.

**Comment by Desvelao** [comment]
# Update
- Add tests

File: server/lib/reporting/printer.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ import * as TimSort from 'timsort';
 import { getConfiguration } from '../get-configuration';
 import { REPORTS_PRIMARY_COLOR} from '../../../common/constants';
 import { getSettingDefaultValue } from '../../../common/services/settings';
+import { try } from 'bluebird';
```
Comment (AlexRuiz7) [inline server/lib/reporting/printer.ts:15]: Check this import

**Comment by Desvelao** [comment]
# Update
- Update development branch with the base branch
- Add test related to the validation of the `customization.reports.footer` and `customization.reports.header` settings in the form inputs

**Comment by Desvelao** [comment]
# Update
- Add test for the `GET /reports` endpoint

**Comment by Desvelao** [comment]
# Update
- Update the development branch with the base branch

File: server/lib/reporting/printer.ts (hunk 1 above)
Comment (Desvelao) [reply server/lib/reporting/printer.ts]: I removed here: https://github.com/wazuh/wazuh-kibana-app/pull/4505/commits/95b9020e4c1d1b238fdcd169532c581fa3524501

File: public/components/common/form/input_text_area.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import {
+	EuiTextArea,
+} from '@elastic/eui';
```
Comment (yenienserrano) [inline public/components/common/form/input_text_area.tsx:4]: It may be possible to put it in 1 line

**Review by yenienserrano — DISMISSED** [review]
LGTM

**Review by Mayons95 — APPROVED** [review]
TR: :heavy_check_mark: 
Test Details

<details>

Screenshot from 2022-10-28 10-23-32

Screenshot from 2022-10-28 10-23-26

Screenshot from 2022-10-28 10-16-06

Screenshot from 2022-10-28 10-15-04

image

</details>

**Review by AlexRuiz7 — CHANGES_REQUESTED** [review]
- Code review: ✔️ 
- Test: 🔴

The customization of the logo, header and footer for the PDFs work quite well:

image


However, I found a few errors:

- The header allows 4 lines of text, however only 3 are actually rendered.
- The inputs limit the number of text lines, but not their length. It's quite easy to break this feature by adding a single, extra long line.

image
image

Texts used:

```
Powered by Wazuh
- This is a custom footer for PDF reports -
```

```
Powered by Wazuh
________________________
- This is a custom header for PDF reports -
- Provided by the @wazuh/frontend-team -
```

**Comment by Desvelao** [comment]
> * Code review: heavy_check_mark
> * Test: red_circle
> 
> The customization of the logo, header and footer for the PDFs work quite well:
> 
> image
> 
> However, I found a few errors:
> 
> * The header allows 4 lines of text, however only 3 are actually rendered.
> * The inputs limit the number of text lines, but not their length. It's quite easy to break this feature by adding a single, extra long line.
> 
> image image
> 
> Texts used:
> 
> ```
> Powered by Wazuh
> - This is a custom footer for PDF reports -
> ```
> 
> ```
> Powered by Wazuh
> ________________________
> - This is a custom header for PDF reports -
> - Provided by the @wazuh/frontend-team -
> ```

Yes, the settings only limit the number of lines and not the length of the lines, we could limit if we consider it.

When I developed it, I considered the length of the lines is under the responsibility of the user, so the plugin has no limit that could affect some use cases. Anyways, we could set some value and check the input is valid.

**Comment by asteriscos** [comment]
#### Added max length per line validation.

Peek 2022-10-28 18-56

**Review by Desvelao — APPROVED** [review]
LGTM!

**Pull request merged**